feat: support down to node 14, support non-nodejs environments

ThaUnknown opened the pull request:
this PR adds support for non-nodeJS environments by removing the explicit undici and fs requirements, the reasoning behind it:
- fetch is globally available in nodeJS, and explicitly importing it covers a very minor range of non-supported minor node versions
- using global fetch allows the user to polyfill it themselves using `node-fetch` allowing the library to run on versions prior to node16, down to node12, example:
```js
import _fetch, {
  Blob as _Blob,
  File as _File,
  FormData as _FormData,
  Headers as _Headers,
  Request as _Request,
  Response as _Response
} from 'node-fetch'

global.fetch = global.fetch || _fetch

global.Blob = global.Blob || _Blob
global.File = global.File || _File
global.FormData = global.FormData || _FormData
global.Headers = global.Headers || _Headers
global.Request = global.Request || _Request
global.Response = global.Response || _Response
```

I also removed the "load by file" functionality, it seems mostly useless, as it's 1 line of code which is easy to write for any developer, and it strictly only allows the module to run on node, removing this allows the module to potentially run in other JS engines than node which implement fetch

Also removed sharp for the same reason, up to the user to crop the image.

dimdenGD commented:
- Global fetch started to be available without flags only in Node.js 20, for anything below you'd need to manually get fetch library and pollute your global scope, which is very not developer-friendly.
- Optional chaining was introduced only in Node.js 14, so at most it'd be down to 14, not 12.
- Load by file is very common thing to do, and it's really nice when library handles it for you, most libraries include such functionality.
- By removing sharp, you make every single developer unknowingly get library half-baked. They will think it works, until they randomly get errors, take time to research into it, and then finally realize what's wrong, then write downsizing themselves, every single developer ever using this library would have to go through this, instead of this functionality already being included in. Also this would break Sharex script for big region captures too.

Basically all of the changes are making this library inconvenient and annoying to use.

ThaUnknown commented:
unpleasant that you closed this instead of letting me address these issues, but okay

dimdenGD commented:
I don't really see possibility for anything to be improved really. In it's current state, I think library would work fine down to Node.js 14 already. And I don't really have interest in supporting other environments (and older versions of Node), afaik most of them support Node.js APIs already (Bun for sure does)

ThaUnknown commented:
it wont as undici is >=Node18, so it simply wont work on older versions, it's costless for you to support non-nodejs enviorments, so I don't understand the argument there

dimdenGD commented:
Doesn't seem to be costless based on changes you did... Also to get support for Node 14, you can just downgrade undici from 6 to 5 (5.28.2 came out 2 months ago)

ThaUnknown commented:
because I did them on a whim without your input.....? you voiced your complaints, and listed out the changes you didn't agree with, but didn't let me address the complaints and closed the PR, there is always more than 1 way of accomplishing something

dimdenGD commented:
I still don't really see possibility for improvement, support for other environments is basically impossible without removing node-specific things like fs, and I don't want that, and I don't want optional chaining to be removed neither. You're free to try to come up with something though, if you want to

ThaUnknown committed:
feat: support down to node 12, support non-nodejs environments

ThaUnknown commented:
I'm not certain I understand the 
> pollute your global scope

if you're using fetch, it already would be globally defined, but I guess you mean that someone would intentionally apply fetch globally? so I fixed this by importing a conditional polyfill?
I think I addressed everything here, you wanted conditional chaining, so the minimum for this is gonna be node14, oh well

I was also uncertain about the code formatting and styling as you seem to have used at least 3 code styles? so I went for some in-between

ThaUnknown committed:
fix: documentation

dimdenGD commented:
I think instead of using different libraries for different node versions it's better to just downgrade undici to previously mentioned version, this way it'll all work down to 14 with no need for different ones

ThaUnknown commented:
undici is 1.2MB in size..... ya sure? you'll be importing a 1.2MB library just for something that's built-in globally

dimdenGD commented:
Yes

dimdenGD commented:
For some reason you removed all semicolons, it shouldn't be like that

ThaUnknown committed:
fix: use undici

ThaUnknown commented:
can't say I agree, oh well

ThaUnknown commented:
> For some reason you removed all semicolons, it shouldn't be like that

because you had some code with semicolons and some without, so I just went with consistency and removed them all

ThaUnknown committed:
chore: semi

dimdenGD committed:
Update README.md

dimdenGD commented:
I believe 'core' would still not work in non-node environments due to use of image-dimensions and set-cookie-parser libraries

ThaUnknown commented:
set-cookie-parser has 0 dependencies, and imports no libraries
same goes for image-dimensions, which only uses JS language functions

dimdenGD commented:
In order for it to work, another environment other than node would need to:
- Support getting library just by it's name from somewhere (node does this from node_modules)
- Support commonjs modules

Regarding first issue, as far as I know, there are 3 major server js environments:
- Node.js: should work fine
- Bun: should work fine, but it was working fine in Bun even before all changes in this PR, so kinda useless
- Deno: doesn't support commonjs modules, so it won't work

ThaUnknown commented:
yes, that's called bundling, again non-issue

dimdenGD commented:
Bundling would only apply to browsers, edited my prev comment

ThaUnknown commented:
again, not really, from the examples you listed it would make it work on deno

and the 4 envs you listed aren't the only ones available, there are many, many, many, many places where this could be used, chrome extensions, userscripts, one of the 10's nativescript based UI libraries, capacitor, cordova, pocketbase, and you can bundle to those targets, or simply bundle to headless targets, bundling isn't exclusive to browsers, for example for deno you could do
```js
import OCR from "https://esm.sh/chrome-lens-ocr/src/core.js?target=deno";
```
and use it on deno, the specific implementation details don't matter, what matters is that it's possible, and exposes that kind of possibility, previously without directly copying and modifying the code it wasn't even possible because of the modules which were moved away from `core`

ThaUnknown commented:
a good example of this are [webpack build targets](https://webpack.js.org/configuration/target/#string), where u can build to specific targets like node, async node, web, web workers, electron, nwjs, or just a specific ecmascript version for any non-standard platform which targets X js version

dimdenGD commented:
Deno doesn't work with `module.exports` which `set-cookie-parser` uses

ThaUnknown commented:
> Deno doesn't work with `module.exports` which `set-cookie-parser` uses

yeeeeeeeeeeeeeees, which is why you use some X bundler to solve that for you! once again, the implementation details don't matter, what matters is that it's possible, in some way or another

ThaUnknown commented:
for example: [esm.sh](https://esm.sh/set-cookie-parser@2.6.0?target=deno%22) automatically converts CJS to ESM for deno

again these are examples, what matters is that importing platform specific modules no longer blocks the core of the library from working on *insert any environment name here*

dimdenGD committed:
Update README.md
